File: game/src/main/java/com/exjets/aetheria/core/Bank.java
package com.exjets.aetheria.core;

import java.util.LinkedHashMap;
import java.util.Map;

/** Unlimited storage, reachable at any bank booth. Every item stacks here. */
public class Bank {

    private final Map<String, Integer> contents = new LinkedHashMap<>();

    public void deposit(String itemId, int count) {
        if (count <= 0) {
            return;
        }
        contents.merge(itemId, count, Integer::sum);
    }

    /** @return how many were actually withdrawn */
    public int withdraw(String itemId, int count) {
        int held = count(itemId);
        int taken = Math.min(held, count);
        if (taken <= 0) {
            return 0;
        }
        if (held - taken <= 0) {
            contents.remove(itemId);
        } else {
            contents.put(itemId, held - taken);
        }
        return taken;
    }

    public int count(String itemId) {
        return contents.getOrDefault(itemId, 0);
    }

    public Map<String, Integer> contents() {
        return contents;
    }

    public boolean isEmpty() {
        return contents.isEmpty();
    }

    public void clear() {
        contents.clear();
    }
}

File: src/game/core/bank.ts
/** Unlimited storage reachable at any bank booth. Everything stacks here. */
export class Bank {
  private readonly contents = new Map<string, number>();

  deposit(id: string, count: number): void {
    if (count <= 0) return;
    this.contents.set(id, this.count(id) + count);
  }

  /** @returns how many were actually withdrawn */
  withdraw(id: string, count: number): number {
    const held = this.count(id);
    const taken = Math.min(held, count);
    if (taken <= 0) return 0;
    if (held - taken <= 0) this.contents.delete(id);
    else this.contents.set(id, held - taken);
    return taken;
  }

  count(id: string): number {
    return this.contents.get(id) ?? 0;
  }

  entries(): [string, number][] {
    return [...this.contents.entries()];
  }

  isEmpty(): boolean {
    return this.contents.size === 0;
  }

  clear(): void {
    this.contents.clear();
  }
}
